### ISSUE TITLE
Questions on Repeat Results from Table I

### ISSUE THREAD

**Issue by BoyuanFeng** [KaidiXu/GCN_ADV_Train#5]
Hi Kaidi,

Thanks for your wonderful work on the PGD attack.

I am trying to repeat your results in Table 1 with the PyTorch code (https://github.com/DSE-MSU/DeepRobust/blob/e6188b04fd8468b0d2349fac3c9f496755cff1e9/deeprobust/graph/global_attack/topology_attack.py).

I have some questions. It would be great if you can kindly share some suggestions and clarifications.

1. What does "fixed natural model" mean in Table 1? Could you please clarify whether it is one of the following guesses?
    a. First guess: We run PGD on the victim model (2-layer GCN) to generate the perturbed topology. Then, we evaluate the accuracy of the victim model on the perturbed topology. Here, we do not retrain the victim model on the perturbed topology and these two models have the same weight.
    b. Second guess: We run PGD on the victim model (2-layer GCN) to generate the perturbed topology. Then, we retrain another 2-layer GCN on the perturbed topology and test on the perturbed topology? Here, these two models have different weights.
    c. If I understand correctly, the paper seems to follow the first guess. However, it seems that the open-source implementation follows the second guess.

2. When I run the code, the accuracy on clean graph is 83.5%, the accuracy on perturbed graph is 75.15% (after 20 samples), which is kind of lower than the reported result of 72% (i.e., the 28% in the Table I for CE-GD on fixed natural model). How to repeat the 28% result in Table I for CE-PGD on fixed natural model?

3. I also implemented my first guess, by directly testing the victim model on the perturbed data without retraining. It produces accuracy of 82.34%. (The accuracy on clean graph is 83.5%). Could you please share some intuition why this happens? Intuitively, the accuracy should drop a lot for the victim model, since this is what we are running the PGD for.

Thanks a lot for all the helps!

**Comment by KaidiXu** [comment]
Hi Boyuan,

Thanks for your interest! 
1. The PyTorch code is not under our maintenance, but for your question, the first guess is correct. The second guess will lead a little bit performance drop.
2. For the CE-PGD attack, 20 steps may not enough. Please use the default setting like https://github.com/KaidiXu/GCN_ADV_Train/blob/7ace67ad2a54804f6a138e52e78ea7fdab34d009/attack.py#L24.
3. This is very wired, I think it could be caused by some bugs.  In our TensorFlow version, the performance is normal, and many people have already reproduced the results based on it.

Best,
Kaidi

**Comment by BoyuanFeng** [comment]
Thanks for your quick response! I will definitely try it.

**Comment by ChandlerBang** [comment]
Hi BoyuanFeng,

I am the provider for PyTorch version of PGD attack. For your questions,
 
1) First I want to note that the data splits provided in DeepRobust [test_pgd.py](https://github.com/DSE-MSU/DeepRobust/blob/master/examples/graph/test_pgd.py) follows the setting of Nettack/Meta-attack, i.e., . 10%/10%/80% for training/validation/test, while the PGD paper uses 20 nodes per class for training. Thus, I believe you should try 
```
data = Dataset(root='/tmp/', name=args.dataset, setting='nettack')
```
instead of 
```
data = Dataset(root='/tmp/', name=args.dataset, setting='gcn')
```
and you can get a much lower accuracy on attacked graphs.

2) I am also having this question on the evasion performance. In the evasion attack setting of meta-attack, you can see meta-attack does not work on evasion setting (the performance drops only 2% see Table 3 in the [paper](https://openreview.net/pdf?id=Bylnx209YX)). I am not sure if it is because the evasion setting in PGD attack paper might be a little different from that in Meta-attack. Or maybe you can try the hyper-parameter setting as Kaidi suggested to see if we can get higher attack success rate. 

Thanks.

**Comment by BoyuanFeng** [comment]
Thanks for your information!

**Comment by ChandlerBang** [comment]
Hi Boyuan,

I have some updates for the evasion performance. We found that the authors also use the pseudo-labels of test data during attack, which is a nice strategy and greatly reduces the evasion accuray. I have updated the [`test_pgd.py`](https://github.com/DSE-MSU/DeepRobust/blob/756453e894df2acd154c0016b9e25836c8960b27/examples/graph/test_pgd.py#L90-L96
) to reflect this.

Feel free to let me know if you have other questions.

**Comment by BoyuanFeng** [comment]
> Hi Boyuan,
> 
> I have some updates for the evasion performance. We found that the authors also use the pseudo-labels of test data during attack, which is a nice strategy and greatly reduces the evasion accuray. I have updated the [`test_pgd.py`](https://github.com/DSE-MSU/DeepRobust/blob/756453e894df2acd154c0016b9e25836c8960b27/examples/graph/test_pgd.py#L90-L96) to reflect this.
> 
> Feel free to let me know if you have other questions.

Thanks for the information!
